subs/pantry/src/Pantry/Types.hs

```diff
@@ -1002,12 +1005,56 @@ parsePackageIdentifier str =
 parsePackageName :: String -> Maybe PackageName
 parsePackageName = Distribution.Text.simpleParse
 
+-- | A package name parse fail.
+data PackageNameParseFail
+  = PackageNameParseFail Text
+  | CabalFileNameParseFail FilePath
+  | CabalFileNameInvalidPackageName FilePath
+  deriving (Typeable)
+instance Exception PackageNameParseFail
+instance Show PackageNameParseFail where
+    show (PackageNameParseFail bs) = "Invalid package name: " ++ show bs
+    show (CabalFileNameParseFail fp) = "Invalid file path for cabal file, must have a .cabal extension: " ++ fp
+    show (CabalFileNameInvalidPackageName fp) = "cabal file names must use valid package names followed by a .cabal extension, the following is invalid: " ++ fp
+
+-- | Parse a package name from a 'String'.
+parsePackageNameThrowing :: MonadThrow m => String -> m PackageName
+parsePackageNameThrowing str =
+  case parsePackageName str of
+    Nothing -> throwM $ PackageNameParseFail $ T.pack str
+    Just pn -> pure pn
+
+-- | Parse a package name from a file path.
+parsePackageNameFromFilePath :: MonadThrow m => Path a File -> m PackageName
```

Comment:
It looks like this function is still be used in one place in the codebase, but I have a feeling it's not really needed. Let's take it out of pantry, and move its implementation directly into `Stack.Package` to avoid cluttering the public API with something that shouldn't in general be needed.

Reply:
Ok, I had some similar thoughts but didn't pursue them, I'll fix it